File: libraries/ecore/src/ecore/impl/AbstractNotification.cpp
// *****************************************************************************
//
// This file is part of a MASA library or program.
// Refer to the included end-user license agreement for restrictions.
//
// Copyright (c) 2018 MASA Group
//
// *****************************************************************************


#include "ecore/impl/AbstractNotification.hpp"
#include "ecore/impl/NotificationChain.hpp"
#include "ecore/Stream.hpp"
#include "ecore/ENotifier.hpp"

using namespace ecore;
using namespace ecore::impl;

bool AbstractNotification::merge( const std::shared_ptr<ENotification>& notification )
{
    switch (eventType_)
    {
        case SET:
        case UNSET:
        {
            EventType notificationEventType = notification->getEventType();
            switch (notificationEventType)
            {
                case SET:
                case UNSET:
                {
                    if (this->getNotifier() == notification->getNotifier() && this->getFeatureID() == notification->getFeatureID())
                    {
                        newValue_ = notification->getNewValue();
                        if (notification->getEventType() == SET)
                            eventType_ = SET;
                        return true;
                    }
                }
            }
        }
        case REMOVE:
        {
            EventType notificationEventType = notification->getEventType();
            switch (notificationEventType)
            {
                case REMOVE:
                {
                    if ( this->getNotifier() == notification->getNotifier() && this->getFeatureID() == notification->getFeatureID())
                    {
                        std::size_t originalPosition = getPosition();
                        std::size_t notificationPosition = notification->getPosition();
                        eventType_ = REMOVE_MANY;
                        std::vector<Any> removedValues;
                        if (originalPosition <= notificationPosition)
                        {
                            removedValues.push_back( oldValue_ );
                            removedValues.push_back( notification->getOldValue() );
                            newValue_ = std::vector<std::size_t>{ position_ = originalPosition, notificationPosition + 1 };
                        }
                        else
                        {
                            removedValues.push_back( notification->getOldValue() );
                            removedValues.push_back( oldValue_ );
                            newValue_ = std::vector<std::size_t>{ position_ = notificationPosition, originalPosition };
                        }
                        oldValue_ = removedValues;
                        return true;
                    }
                    break;
                }
            }
            break;
        }
        case REMOVE_MANY:
        {
            EventType notificationEventType = notification->getEventType();
            switch (notificationEventType)
            {
                case REMOVE:
                {
                    if ( this->getNotifier() == notification->getNotifier() && this->getFeatureID() == notification->getFeatureID())
                    {
                        std::size_t notificationPosition = notification->getPosition();
                        std::vector<std::size_t> positions = std::move( anyCast<std::vector<std::size_t>>(newValue_) );
                        std::vector<std::size_t> newPositions( positions.size() + 1 );

                        int index = 0;
                        while (index < positions.size())
                        {
                            std::size_t oldPosition = positions[index];
                            if (oldPosition <= notificationPosition)
                            {
                                newPositions[index++] = oldPosition;
                                ++notificationPosition;
                            }
                            else
                                break;
                        }

                        std::vector<Any> oldValue = std::move( anyCast<std::vector<Any>>(oldValue_) );
                        oldValue.insert( oldValue.begin() + index, notification->getOldValue() );

                        newPositions[index] = notificationPosition;
                        while (++index < positions.size() + 1)
                            newPositions[index] = positions[index - 1];

                        oldValue_ = oldValue;
                        newValue_ = newPositions;
                        return true;
                    }
                    break;
                }
            }
            break;
        }
    }
    return false;
}

bool AbstractNotification::add( const std::shared_ptr<ENotification>& notification )
{
    if (!notification)
        return false;
    if (merge( notification ))
        return false;
    if (!next_)
    {
        if (auto n = std::dynamic_pointer_cast<AbstractNotification>(notification))
        {
            next_ = n;
            return true;
        }
        else
        {
            next_ = std::make_shared<NotificationChain>();
            return next_->add( notification );
        }
    }
    else
        return next_->add( notification );
}

void AbstractNotification::dispatch()
{
    auto notifier = getNotifier();
    if ( notifier && eventType_ != -1)
        notifier->eNotify( std::enable_shared_from_this<AbstractNotification>::shared_from_this() );
    if (next_)
        next_->dispatch();
}

File: ecore/go/src/pkg/ecore/abstractnotification.go
package ecore

type AbstractNotification struct {
	ENotification
	eventType_ EventType
	oldValue_  interface{}
	newValue_  interface{}
	position_  int
	next_      ENotificationChain
}

func (notif *AbstractNotification) GetEventType() EventType {
	return notif.eventType_
}

func (notif *AbstractNotification) GetOldValue() interface{} {
	return notif.oldValue_
}

func (notif *AbstractNotification) GetNewValue() interface{} {
	return notif.newValue_
}

func (notif *AbstractNotification) GetPosition() int {
	return notif.position_
}

func (notif *AbstractNotification) Merge(eNotif ENotification) bool {
	switch ev := notif.eventType_; ev {
	case SET:
	case UNSET:
		switch notifEv := eNotif.GetEventType(); notifEv {
		case SET:
		case UNSET:
			if notif.ENotification.GetNotifier() == eNotif.GetNotifier() &&
				notif.ENotification.GetFeatureId() == eNotif.GetFeatureId() {
				notif.newValue_ = eNotif.GetNewValue()
				if notif.GetEventType() == SET {
					notif.eventType_ = SET
					return true
				}
			}
		}
	case REMOVE:
		switch notifEv := eNotif.GetEventType(); notifEv {
		case REMOVE:
			if notif.ENotification.GetNotifier() == eNotif.GetNotifier() &&
				notif.ENotification.GetFeatureId() == eNotif.GetFeatureId() {
				originalPosition := notif.ENotification.GetPosition()
				notificationPosition := eNotif.GetPosition()
				notif.eventType_ = REMOVE_MANY
				removedValues := NewArrayEList([]interface{}{})
				if originalPosition <= notificationPosition {
					removedValues.Add(notif.oldValue_)
					removedValues.Add(eNotif.GetOldValue())
					notif.position_ = originalPosition
					notif.newValue_ = NewArrayEList([]interface{}{originalPosition, notificationPosition + 1})
				} else {
					removedValues.Add(eNotif.GetOldValue())
					removedValues.Add(notif.oldValue_)
					notif.position_ = notificationPosition
					notif.newValue_ = NewArrayEList([]interface{}{notificationPosition, originalPosition})
				}
				notif.oldValue_ = removedValues
				return true
			}
		}
	case REMOVE_MANY:
		switch notifEv := eNotif.GetEventType(); notifEv {
		case REMOVE:
			if notif.ENotification.GetNotifier() == eNotif.GetNotifier() &&
				notif.ENotification.GetFeatureId() == eNotif.GetFeatureId() {
				notificationPosition := eNotif.GetPosition()
				positions := notif.newValue_.(EList)
				newPositions := NewArrayEList([]interface{}{})
				for i := 0; i < positions.Size()+1; i++ {
					newPositions.Add(0)
				}
				index := 0
				for index < positions.Size() {
					oldPosition := positions.Get(index).(int)
					if oldPosition <= notificationPosition {
						newPositions.Set(index, oldPosition)
						index++
					} else {
						break
					}
				}
				notif.oldValue_.(EList).Insert(index, eNotif.GetOldValue())
				newPositions.Set(index, notificationPosition)
				index++
				for ; index < positions.Size()+1; index++ {
					newPositions.Set(index, positions.Get(index-1))
				}
				notif.newValue_ = newPositions
				return true
			}
		}
	}
	return false
}

func (notif *AbstractNotification) Add(eNotif ENotification) bool {
	if eNotif == nil {
		return false
	}
	if notif.Merge(eNotif) {
		return false
	}
	if notif.next_ == nil {
		value, ok := eNotif.(ENotificationChain)
		if ok {
			notif.next_ = value
			return true
		} else {
			notif.next_ = &NotificationChain{}
			return notif.next_.Add(eNotif)
		}
	} else {
		return notif.next_.Add(eNotif)
	}
}

func (notif *AbstractNotification) Dispatch() {
	notifier := notif.ENotification.GetNotifier()
	if notifier != nil && notif.eventType_ != -1 {
		notifier.ENotify(notif)
	}
	if notif.next_ != nil {
		notif.next_.Dispatch()
	}
}
